File: duckdb/duckdb.go
// Package duckdb provides native DuckDB integration for Solod.
//
// This implementation ships inside [soloduck] so applications can link libduckdb while using
// stock solod.dev from [upstream Solod]—the core repo does not ship this package.
//
// This package is designed around DuckDB's C API with explicit resource
// ownership and predictable lifecycles. Query execution follows duckdb_query
// semantics: after any call that fills a result object, Result.Close must run
// even when the query fails, so errors attached to the result are released correctly.
//
// Follow the DuckDB C overview flow (install libduckdb, use duckdb.h): open a database,
// connect, run queries, disconnect and close. See:
//   - https://duckdb.org/docs/current/clients/c/overview
//   - https://duckdb.org/docs/current/clients/c/connect
//
// Query/result details align with:
// https://duckdb.org/docs/current/clients/c/query
//
// Building translated programs that import this package requires DuckDB headers
// and library to be available to the C compiler and linker (e.g. `-lduckdb`).
//
// [soloduck]: https://github.com/lmangani/soloduck
// [upstream Solod]: https://github.com/solod-dev/solod
package duckdb

import (
	"solod.dev/so/c"
	"solod.dev/so/errors"
	"solod.dev/so/mem"
	"solod.dev/so/strings"
)

var (
	ErrOpen       = errors.New("duckdb: open failed")
	ErrClosed     = errors.New("duckdb: connection is closed")
	ErrQuery      = errors.New("duckdb: query failed")
	ErrPrepare    = errors.New("duckdb: prepare failed")
	ErrBind       = errors.New("duckdb: bind failed")
	ErrExec       = errors.New("duckdb: execute failed")
	ErrNullValue  = errors.New("duckdb: null value")
	ErrInvalidCol = errors.New("duckdb: invalid column index")
	ErrInvalidRow = errors.New("duckdb: invalid row index")
)

// Conn is a connection to a DuckDB database instance.
type Conn struct {
	db     dbHandle
	closed bool
}

// ConnResult is a helper struct for returning
// a Conn and an error from a function.
type ConnResult struct {
	val Conn
	err error
}

// Stmt is a prepared statement bound to a connection.
type Stmt struct {
	stmt   stmtHandle
	closed bool
}

// StmtResult is a helper struct for returning
// a Stmt and an error from a function.
type StmtResult struct {
	val Stmt
	err error
}

// Result owns a DuckDB query result.
type Result struct {
	res    resultHandle
	closed bool
}

// ResultResult is a helper struct for returning
// a Result and an error from a function.
type ResultResult struct {
	val Result
	err error
}

// Rows is a row iterator over a Result.
type Rows struct {
	result *Result
	row    int
}

// LibraryVersion returns duckdb_library_version() (e.g. "v1.5.2").
// The string is owned by DuckDB; do not free it.
func LibraryVersion() string {
	p := so_duckdb_library_version()
	if p == nil {
		return ""
	}
	return c.String(p)
}

// Open creates a new DuckDB database handle and connection from a path string.
// Use a file path for an on-disk database or ":memory:" for an in-memory database string path.
//
// For the same in-memory setup as the C examples using duckdb_open(NULL), use [OpenInMemory].
func Open(path string) (Conn, error) {
	var db dbHandle
	if so_duckdb_open(path, &db) != 0 {
		return Conn{}, ErrOpen
	}
	return Conn{db: db}, nil
}

// OpenInMemory opens an ephemeral database using duckdb_open(NULL) and duckdb_connect,
// matching the Startup & Shutdown C examples.
func OpenInMemory() (Conn, error) {
	var db dbHandle
	if so_duckdb_open_memory(&db) != 0 {
		return Conn{}, ErrOpen
	}
	return Conn{db: db}, nil
}

// Close closes the connection and frees underlying DuckDB resources.
func (c *Conn) Close() error {
	if c.closed {
		return nil
	}
	so_duckdb_close(&c.db)
	c.closed = true
	return nil
}

// Interrupt requests interruption of the current query on this connection (duckdb_interrupt).
func (c *Conn) Interrupt() {
	if c.closed {
		return
	}
	so_duckdb_interrupt(&c.db)
}

// Query executes SQL and returns a materialized result set.
//
// If the returned error is non-nil, the [Result] still wraps the DuckDB result
// object (including error details); the caller must call [Result.Close].
func (c *Conn) Query(query string) (Result, error) {
	if c.closed {
		return Result{}, ErrClosed
	}
	var res resultHandle
	rc := so_duckdb_query(&c.db, query, &res)
	out := Result{res: res}
	if rc != 0 {
		return out, ErrQuery
	}
	return out, nil
}

// Exec executes SQL and returns the number of changed rows.
func (c *Conn) Exec(query string) (int, error) {
	res, err := c.Query(query)
	if err != nil {
		_ = res.Close()
		return 0, err
	}
	n := res.RowsChanged()
	closeErr := res.Close()
	return n, closeErr
}

// ExecSQL runs SQL without retaining a result set (duckdb_query with a NULL result pointer).
// Use this for DDL or statements where row metadata is not needed.
//
// Errors return [ErrQuery] without an error message; use [Conn.Query] when diagnostics are required.
func (c *Conn) ExecSQL(sql string) error {
	if c.closed {
		return ErrClosed
	}
	if so_duckdb_query_void(&c.db, sql) != 0 {
		return ErrQuery
	}
	return nil
}

// Prepare creates a prepared statement.
func (c *Conn) Prepare(query string) (Stmt, error) {
	if c.closed {
		return Stmt{}, ErrClosed
	}
	var stmt stmtHandle
	if so_duckdb_prepare(&c.db, query, &stmt) != 0 {
		return Stmt{}, ErrPrepare
	}
	return Stmt{stmt: stmt}, nil
}

// Close releases prepared statement resources.
func (s *Stmt) Close() error {
	if s.closed {
		return nil
	}
	so_duckdb_stmt_close(&s.stmt)
	s.closed = true
	return nil
}

// ClearBindings clears all existing parameter bindings.
func (s *Stmt) ClearBindings() error {
	if s.closed {
		return ErrClosed
	}
	if so_duckdb_stmt_clear(&s.stmt) != 0 {
		return ErrBind
	}
	return nil
}

// BindNull binds NULL to parameter index (1-based).
func (s *Stmt) BindNull(index int) error {
	if s.closed {
		return ErrClosed
	}
	if so_duckdb_bind_null(&s.stmt, index) != 0 {
		return ErrBind
	}
	return nil
}

// BindBool binds a bool value to parameter index (1-based).
func (s *Stmt) BindBool(index int, value bool) error {
	if s.closed {
		return ErrClosed
	}
	if so_duckdb_bind_bool(&s.stmt, index, value) != 0 {
		return ErrBind
	}
	return nil
}

// BindInt64 binds an int64 value to parameter index (1-based).
func (s *Stmt) BindInt64(index int, value int64) error {
	if s.closed {
		return ErrClosed
	}
	if so_duckdb_bind_int64(&s.stmt, index, value) != 0 {
		return ErrBind
	}
	return nil
}

// BindFloat64 binds a float64 value to parameter index (1-based).
func (s *Stmt) BindFloat64(index int, value float64) error {
	if s.closed {
		return ErrClosed
	}
	if so_duckdb_bind_double(&s.stmt, index, value) != 0 {
		return ErrBind
	}
	return nil
}

// BindString binds a string value to parameter index (1-based).
func (s *Stmt) BindString(index int, value string) error {
	if s.closed {
		return ErrClosed
	}
	if so_duckdb_bind_varchar(&s.stmt, index, value) != 0 {
		return ErrBind
	}
	return nil
}

// Query executes the prepared statement and returns a result set.
//
// If the returned error is non-nil, the [Result] still must be closed.
func (s *Stmt) Query() (Result, error) {
	if s.closed {
		return Result{}, ErrClosed
	}
	var res resultHandle
	rc := so_duckdb_stmt_exec(&s.stmt, &res)
	out := Result{res: res}
	if rc != 0 {
		return out, ErrExec
	}
	return out, nil
}

// Exec executes the prepared statement and returns changed row count.
func (s *Stmt) Exec() (int, error) {
	res, err := s.Query()
	if err != nil {
		_ = res.Close()
		return 0, err
	}
	n := res.RowsChanged()
	closeErr := res.Close()
	return n, closeErr
}

// PrepareError returns the latest prepare error for this statement.
func (s *Stmt) PrepareError() string {
	msg := so_duckdb_prepare_error(&s.stmt)
	if msg == nil {
		return ""
	}
	return c.String(msg)
}

// Close releases the underlying result data.
func (r *Result) Close() error {
	if r.closed {
		return nil
	}
	so_duckdb_result_close(&r.res)
	r.closed = true
	return nil
}

// Error returns the result-level error message, if any.
//
// The pointer is owned by DuckDB; do not free it. It is valid until [Result.Close].
func (r *Result) Error() string {
	msg := so_duckdb_result_error(&r.res)
	if msg == nil {
		return ""
	}
	return c.String(msg)
}

// ErrorType returns the error classification when [Conn.Query] failed.
func (r *Result) ErrorType() ErrorType {
	if r.closed {
		return ErrorInvalid
	}
	return ErrorType(so_duckdb_result_error_type(&r.res))
}

// StatementType returns the statement type that produced this result.
func (r *Result) StatementType() StatementType {
	if r.closed {
		return StatementInvalid
	}
	return StatementType(so_duckdb_result_statement_type(&r.res))
}

// RowCount returns the number of rows in this result.
func (r *Result) RowCount() int {
	return so_duckdb_result_row_count(&r.res)
}

// RowsChanged returns number of rows changed by the statement.
func (r *Result) RowsChanged() int {
	return so_duckdb_result_rows_changed(&r.res)
}

// ColumnCount returns number of columns in this result.
func (r *Result) ColumnCount() int {
	return so_duckdb_result_column_count(&r.res)
}

// ColumnName returns a column name by index.
func (r *Result) ColumnName(col int) (string, error) {
	if col < 0 || col >= r.ColumnCount() {
		return "", ErrInvalidCol
	}
	name := so_duckdb_result_column_name(&r.res, col)
	if name == nil {
		return "", ErrInvalidCol
	}
	return c.String(name), nil
}

// ColumnType returns the physical SQL type of column col (see [ColType] constants).
func (r *Result) ColumnType(col int) (ColType, error) {
	if r.closed {
		return ColInvalid, ErrQuery
	}
	if col < 0 || col >= r.ColumnCount() {
		return ColInvalid, ErrInvalidCol
	}
	return ColType(so_duckdb_column_type(&r.res, col)), nil
}

// ColumnData returns duckdb_column_data: a pointer to columnar data for col.
// Layout depends on [Result.ColumnType] / [ColType]; see DuckDB C documentation.
func (r *Result) ColumnData(col int) any {
	if r.closed || col < 0 || col >= r.ColumnCount() {
		return nil
	}
	return so_duckdb_column_data(&r.res, col)
}

// NullmaskData returns duckdb_nullmask_data for col, or nil if unavailable.
func (r *Result) NullmaskData(col int) any {
	if r.closed || col < 0 || col >= r.ColumnCount() {
		return nil
	}
	return so_duckdb_nullmask_data(&r.res, col)
}

// ColumnLogicalType returns an opaque duckdb_logical_type handle for col (see DuckDB C docs).
// Release it with [DestroyLogicalType].
func (r *Result) ColumnLogicalType(col int) (any, error) {
	if r.closed {
		return nil, ErrQuery
	}
	if col < 0 || col >= r.ColumnCount() {
		return nil, ErrInvalidCol
	}
	h := so_duckdb_column_logical_type(&r.res, col)
	if h == nil {
		return nil, ErrInvalidCol
	}
	return h, nil
}

// DestroyLogicalType destroys a handle returned by [Result.ColumnLogicalType].
func DestroyLogicalType(lt any) {
	if lt == nil {
		return
	}
	so_duckdb_logical_type_destroy(lt)
}

// IsNull reports whether a value at (row, col) is null.
func (r *Result) IsNull(row int, col int) (bool, error) {
	if row < 0 || row >= r.RowCount() {
		return false, ErrInvalidRow
	}
	if col < 0 || col >= r.ColumnCount() {
		return false, ErrInvalidCol
	}
	return so_duckdb_value_is_null(&r.res, col, row), nil
}

// Bool returns a bool value from (row, col).
func (r *Result) Bool(row int, col int) (bool, error) {
	isNull, err := r.IsNull(row, col)
	if err != nil {
		return false, err
	}
	if isNull {
		return false, ErrNullValue
	}
	return so_duckdb_value_bool(&r.res, col, row), nil
}

// Int64 returns an int64 value from (row, col).
func (r *Result) Int64(row int, col int) (int64, error) {
	isNull, err := r.IsNull(row, col)
	if err != nil {
		return 0, err
	}
	if isNull {
		return 0, ErrNullValue
	}
	return so_duckdb_value_int64(&r.res, col, row), nil
}

// Float64 returns a float64 value from (row, col).
func (r *Result) Float64(row int, col int) (float64, error) {
	isNull, err := r.IsNull(row, col)
	if err != nil {
		return 0, err
	}
	if isNull {
		return 0, ErrNullValue
	}
	return so_duckdb_value_double(&r.res, col, row), nil
}

// StringCopy returns a heap-allocated copy of a string value from (row, col).
// The caller owns the returned string and must free it using mem.FreeString.
func (r *Result) StringCopy(a mem.Allocator, row int, col int) (string, error) {
	isNull, err := r.IsNull(row, col)
	if err != nil {
		return "", err
	}
	if isNull {
		return "", ErrNullValue
	}
	ptr := so_duckdb_value_string(&r.res, col, row)
	if ptr == nil {
		return "", ErrQuery
	}
	tmp := c.String(ptr)
	out := strings.Clone(a, tmp)
	so_duckdb_string_free(ptr)
	return out, nil
}

// Rows creates a row iterator over this result.
func (r *Result) Rows() Rows {
	return Rows{
		result: r,
		row:    -1,
	}
}

// Next advances to the next row and reports whether one exists.
func (rs *Rows) Next() bool {
	if rs.result == nil {
		return false
	}
	rs.row++
	return rs.row < rs.result.RowCount()
}

// Row returns the current row index.
func (rs *Rows) Row() int {
	return rs.row
}

// IsNull reports whether current row value at col is null.
func (rs *Rows) IsNull(col int) (bool, error) {
	if rs.result == nil {
		return false, ErrQuery
	}
	return rs.result.IsNull(rs.row, col)
}

// Bool reads a bool from current row at col.
func (rs *Rows) Bool(col int) (bool, error) {
	if rs.result == nil {
		return false, ErrQuery
	}
	return rs.result.Bool(rs.row, col)
}

// Int64 reads an int64 from current row at col.
func (rs *Rows) Int64(col int) (int64, error) {
	if rs.result == nil {
		return 0, ErrQuery
	}
	return rs.result.Int64(rs.row, col)
}

// Float64 reads a float64 from current row at col.
func (rs *Rows) Float64(col int) (float64, error) {
	if rs.result == nil {
		return 0, ErrQuery
	}
	return rs.result.Float64(rs.row, col)
}

// StringCopy reads a string from current row at col and clones it.
// The caller owns the returned string and must free it using mem.FreeString.
func (rs *Rows) StringCopy(a mem.Allocator, col int) (string, error) {
	if rs.result == nil {
		return "", ErrQuery
	}
	return rs.result.StringCopy(a, rs.row, col)
}

File: duckdb/duckdb.c
//go:build ignore
#include "duckdb.h"

static int so_duckdb_open_impl(const char* path, so_duckdb_db* out) {
    if (!out) {
        return 1;
    }
    out->db = NULL;
    out->conn = NULL;
    out->open = false;

    if (duckdb_open(path, &out->db) != DuckDBSuccess) {
        return 1;
    }
    if (duckdb_connect(out->db, &out->conn) != DuckDBSuccess) {
        duckdb_close(&out->db);
        return 1;
    }
    out->open = true;
    return 0;
}

int so_duckdb_open(const char* path, so_duckdb_db* out) {
    return so_duckdb_open_impl(path, out);
}

int so_duckdb_open_memory(so_duckdb_db* out) {
    return so_duckdb_open_impl(NULL, out);
}

void so_duckdb_close(so_duckdb_db* db) {
    if (!db || !db->open) {
        return;
    }
    duckdb_disconnect(&db->conn);
    duckdb_close(&db->db);
    db->open = false;
}

void so_duckdb_interrupt(so_duckdb_db* db) {
    if (!db || !db->open) {
        return;
    }
    duckdb_interrupt(db->conn);
}

const char* so_duckdb_library_version(void) {
    return duckdb_library_version();
}

int so_duckdb_query(so_duckdb_db* db, const char* query, so_duckdb_result* out) {
    if (!db || !db->open || !out) {
        return 1;
    }
    duckdb_state st = duckdb_query(db->conn, query, &out->result);
    out->open = true;
    return st == DuckDBSuccess ? 0 : 1;
}

int so_duckdb_query_void(so_duckdb_db* db, const char* query) {
    if (!db || !db->open) {
        return 1;
    }
    return duckdb_query(db->conn, query, NULL) == DuckDBSuccess ? 0 : 1;
}

int so_duckdb_prepare(so_duckdb_db* db, const char* query, so_duckdb_stmt* out) {
    if (!db || !db->open || !out) {
        return 1;
    }
    out->stmt = NULL;
    out->open = false;
    if (duckdb_prepare(db->conn, query, &out->stmt) != DuckDBSuccess) {
        return 1;
    }
    out->open = true;
    return 0;
}

const char* so_duckdb_prepare_error(so_duckdb_stmt* stmt) {
    if (!stmt || !stmt->open) {
        return NULL;
    }
    return duckdb_prepare_error(stmt->stmt);
}

void so_duckdb_stmt_close(so_duckdb_stmt* stmt) {
    if (!stmt || !stmt->open) {
        return;
    }
    duckdb_destroy_prepare(&stmt->stmt);
    stmt->open = false;
}

int so_duckdb_stmt_clear(so_duckdb_stmt* stmt) {
    if (!stmt || !stmt->open) {
        return 1;
    }
    return duckdb_clear_bindings(stmt->stmt) == DuckDBSuccess ? 0 : 1;
}

int so_duckdb_bind_null(so_duckdb_stmt* stmt, int index) {
    if (!stmt || !stmt->open || index <= 0) {
        return 1;
    }
    return duckdb_bind_null(stmt->stmt, (idx_t)index) == DuckDBSuccess ? 0 : 1;
}

int so_duckdb_bind_bool(so_duckdb_stmt* stmt, int index, bool value) {
    if (!stmt || !stmt->open || index <= 0) {
        return 1;
    }
    return duckdb_bind_boolean(stmt->stmt, (idx_t)index, value) == DuckDBSuccess ? 0 : 1;
}

int so_duckdb_bind_int64(so_duckdb_stmt* stmt, int index, int64_t value) {
    if (!stmt || !stmt->open || index <= 0) {
        return 1;
    }
    return duckdb_bind_int64(stmt->stmt, (idx_t)index, value) == DuckDBSuccess ? 0 : 1;
}

int so_duckdb_bind_double(so_duckdb_stmt* stmt, int index, double value) {
    if (!stmt || !stmt->open || index <= 0) {
        return 1;
    }
    return duckdb_bind_double(stmt->stmt, (idx_t)index, value) == DuckDBSuccess ? 0 : 1;
}

int so_duckdb_bind_varchar(so_duckdb_stmt* stmt, int index, const char* value) {
    if (!stmt || !stmt->open || index <= 0) {
        return 1;
    }
    return duckdb_bind_varchar(stmt->stmt, (idx_t)index, value) == DuckDBSuccess ? 0 : 1;
}

int so_duckdb_stmt_exec(so_duckdb_stmt* stmt, so_duckdb_result* out) {
    if (!stmt || !stmt->open || !out) {
        return 1;
    }
    duckdb_state st = duckdb_execute_prepared(stmt->stmt, &out->result);
    out->open = true;
    return st == DuckDBSuccess ? 0 : 1;
}

void so_duckdb_result_close(so_duckdb_result* res) {
    if (!res || !res->open) {
        return;
    }
    duckdb_destroy_result(&res->result);
    res->open = false;
}

const char* so_duckdb_result_error(so_duckdb_result* res) {
    if (!res || !res->open) {
        return NULL;
    }
    return duckdb_result_error(&res->result);
}

int32_t so_duckdb_result_error_type(so_duckdb_result* res) {
    if (!res || !res->open) {
        return 0;
    }
    return (int32_t)duckdb_result_error_type(&res->result);
}

int32_t so_duckdb_result_statement_type(so_duckdb_result* res) {
    if (!res || !res->open) {
        return 0;
    }
    return (int32_t)duckdb_result_statement_type(res->result);
}

int so_duckdb_result_row_count(so_duckdb_result* res) {
    if (!res || !res->open) {
        return 0;
    }
    return (int)duckdb_row_count(&res->result);
}

int so_duckdb_result_rows_changed(so_duckdb_result* res) {
    if (!res || !res->open) {
        return 0;
    }
    return (int)duckdb_rows_changed(&res->result);
}

int so_duckdb_result_column_count(so_duckdb_result* res) {
    if (!res || !res->open) {
        return 0;
    }
    return (int)duckdb_column_count(&res->result);
}

const char* so_duckdb_result_column_name(so_duckdb_result* res, int col) {
    if (!res || !res->open || col < 0) {
        return NULL;
    }
    return duckdb_column_name(&res->result, (idx_t)col);
}

int32_t so_duckdb_column_type(so_duckdb_result* res, int col) {
    if (!res || !res->open || col < 0) {
        return (int32_t)DUCKDB_TYPE_INVALID;
    }
    return (int32_t)duckdb_column_type(&res->result, (idx_t)col);
}

void* so_duckdb_column_data(so_duckdb_result* res, int col) {
    if (!res || !res->open || col < 0) {
        return NULL;
    }
    return duckdb_column_data(&res->result, (idx_t)col);
}

bool* so_duckdb_nullmask_data(so_duckdb_result* res, int col) {
    if (!res || !res->open || col < 0) {
        return NULL;
    }
    return duckdb_nullmask_data(&res->result, (idx_t)col);
}

duckdb_logical_type so_duckdb_column_logical_type(so_duckdb_result* res, int col) {
    if (!res || !res->open || col < 0) {
        return NULL;
    }
    return duckdb_column_logical_type(&res->result, (idx_t)col);
}

void so_duckdb_logical_type_destroy(duckdb_logical_type lt) {
    duckdb_logical_type local = lt;
    duckdb_destroy_logical_type(&local);
}

bool so_duckdb_value_is_null(so_duckdb_result* res, int col, int row) {
    if (!res || !res->open || col < 0 || row < 0) {
        return true;
    }
    return duckdb_value_is_null(&res->result, (idx_t)col, (idx_t)row);
}

bool so_duckdb_value_bool(so_duckdb_result* res, int col, int row) {
    return duckdb_value_boolean(&res->result, (idx_t)col, (idx_t)row);
}

int64_t so_duckdb_value_int64(so_duckdb_result* res, int col, int row) {
    return duckdb_value_int64(&res->result, (idx_t)col, (idx_t)row);
}

double so_duckdb_value_double(so_duckdb_result* res, int col, int row) {
    return duckdb_value_double(&res->result, (idx_t)col, (idx_t)row);
}

char* so_duckdb_value_string(so_duckdb_result* res, int col, int row) {
    return duckdb_value_varchar(&res->result, (idx_t)col, (idx_t)row);
}

void so_duckdb_string_free(char* ptr) {
    if (ptr) {
        duckdb_free(ptr);
    }
}
